src/topology/config/component.rs

```diff
@@ -0,0 +1,113 @@
+use inventory;
+use serde::de::{Deserializer, Error};
+use serde::ser::Serializer;
+use serde::{Deserialize, Serialize};
+use toml::Value;
+
+/// Combines a type field and a nested plugin config to create a table.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ConfigSwapOut {
+    #[serde(rename = "type")]
+    pub type_str: String,
+    #[serde(flatten)]
+    pub nested: Value,
+}
+
+/// Stores both a constructed plugin instance and the ConfigSwapOut used to
+/// create it, this allows us to echo back the config of a component at runtime.
+pub struct ComponentConfig<T>
+where
+    T: 'static + Sized,
+    inventory::iter<ComponentBuilder<T>>:
+        std::iter::IntoIterator<Item = &'static ComponentBuilder<T>>,
+{
+    swap_out: ConfigSwapOut,
+
+    pub component: T,
+}
+
+impl<T> ComponentConfig<T>
+where
+    T: 'static + Sized,
+    inventory::iter<ComponentBuilder<T>>:
+        std::iter::IntoIterator<Item = &'static ComponentBuilder<T>>,
+{
+    /// Returns a sorted Vec of all plugins registered of a type.
+    pub fn types() -> Vec<&'static str> {
+        let mut types = Vec::new();
+        for definition in inventory::iter::<ComponentBuilder<T>> {
+            types.push(definition.name);
+        }
+        types.sort();
+        types
+    }
+}
+
+impl<T> Serialize for ComponentConfig<T>
```

Comment:
Why do we need `Serialize` and `Deserialize` for `ComponentConfig`? It feels like it'd be cleaner to drop these impls and just serialize/deserialize the config itself when needed.

Reply:
We could do, my goal was to hide away the conversion to/from `toml::Value` so we don't have to write it out for each component type, but if it's not saving us much effort then I'm happy to do away with it.

I think it makes sense for me to make a complementary branch that demonstrates converting one of the existing component types to this method. That'll give us a chance to see it in action and decide if we need `ComponentConfig` at all.